File: daily/Stack/503-Next-greater-element-ii.go
package daily

func nextGreaterElements(nums []int) []int {
	n := len(nums)
	stacks := []int{}
	ans := make([]int, n)

	for i := 0; i < n; i++ {
		ans[i] = -1
	}

	for i := 0; i < 2*n; i++ {
		currentNum := nums[i%n]

		for len(stacks) > 0 && nums[stacks[len(stacks)-1]] < currentNum {
			ans[stacks[len(stacks)-1]] = currentNum
			stacks = stacks[:len(stacks)-1]
		}
		if i < n {
			stacks = append(stacks, i)
		}
	}

	return ans
}

File: daily/Stack/503-Next-greater-element-ii.py
class Solution:
    def nextGreaterElements(self, nums: List[int]) -> List[int]:
        n = len(nums)
        stacks = []
        ans = [-1] * n
        for i in range(2 * n):
            current_num = nums[i % n]

            while stacks and nums[stacks[-1]] < current_num:
                ans[stacks[-1]] = current_num
                stacks.pop()
            if i < n:
                stacks.append(i)
        
        return ans